### PR TITLE
HERD NuclearCase MultiPeriod Update

### LINKED ISSUE #245: [TASK] Update HERD to be compatible with newest DISPATCHES release
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
The newest DISPATCHES release upgrades to the newest IDAES version which adds functionality to the MultiPeriod class (it is now more compatible with stochastic runs). Within HERON, HERD should be able to use the MultiPeriod class when running the nuclear case study, just like the nuclear case study notebook in DISPATCHES. DISPATCHES also did some directory reorganizing, so paths need to be updated.

**Describe the solution you'd like**
HERD should build the MultiPeriod class and use it to create the Pyomo model. HERD should link functions in the init call of the MultiPeriod for fixing DoF, flowsheets, etc. 

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #248: [TASK] NE Case Study in HERD could be faster by cutting out redundant build steps
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
Building the Multiperiod model for the nuclear energy case study in HERD takes a while. 

**Describe the solution you'd like**
In the DISPATCHES repo, the renewables case study has a way of speeding up construction of the full model. Instead of calling the flowsheet directly, it calls an intermediate function (e.g., "wind_battery_pem_tank_turb_mp_block" from [here](https://github.com/gmlc-dispatches/dispatches/blob/main/dispatches/case_studies/renewables_case/wind_battery_PEM_tank_turbine_LMP.py)) that calls on the flowsheet once to create the model and on subsequent calls will clone the original model. It then proceeds to propagate any states using data pertaining to that timestep. 

We can implement something similar for the nuclear case study in HERD to speed up the build process and the overall run time. It would also bring HERD closer to a more standardized structure (overall goal being to run any flowsheet in a generalized way). The intermediate function in DISPATCHES is made specifically for one case study, so HERD would need to create a more generalizable intermediate function instead. 

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #250: [DEFECT] HERD amplifies NPV results when running more than 1 scenario
--------
Defect Description
--------

**Describe the defect**

##### What did you expect to see happen?
When running HERD using the `nuclearCase_notebook_match` using more than 1 scenario/sampling, the NPV solution should be within a reasonable range of the the solution for 1 scenario and match results from the [DISPATCHES jupyter notebook](https://github.com/gmlc-dispatches/dispatches/blob/8226a8910664979236afee520ded48062a73b1cc/dispatches/case_studies/nuclear_case/multiperiod_design_pricetaker.ipynb).
Specifically, the results for 1 and 2 samples in the DISPATCHES repository nuclear case notebook are:

N=1,  1.59676063340206e9
N=2,  1.63105178396091e9

##### What did you see instead?
The results from HERD `nuclearCase_notebook_match` are amplified when `num_arma_samples` > 1:

N=1, 1.59402411407554e9
N=2, 3.26420263603555e9

Note: difference between both N=1 results is fine

##### Do you have a suggested fix for the development team?
When creating the TEAL metrics, I believe something is being duplicated in HERD. Perhaps split the metrics (e.g., NPV) per scenario.

**Describe how to Reproduce**
1. Change `<num_arma_samples>1</num_arma_samples>` from 1 to any other integer in the `nuclearCase_notebook_match` test.

**Screenshots and Input Files**
Input file: 

**Platform (please complete the following information):**
 - OS: Windows
 - Versions: Happens in current HERON, but below are relevant tags to recreate error. 
         - HERON: [relevant tag in HERON fork](https://github.com/GabrielSoto-INL/HERON/releases/tag/first_NE_notebook_match)
         - RAVEN: [old commit that matches ROM usage in HERON example](https://github.com/idaholab/raven/commit/e04413c5559a317e8b214de66896578d40b0edde)
 - Dependencies Installation: dispatches=1.0.0 but also occurs in dispatches=1.1.0 (error is not due to this package)

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [ ] 1. Is it tagged with a type: defect or task?
- [ ] 2. Is it tagged with a priority: critical, normal or minor?
- [ ] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [ ] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [ ] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by GabrielSoto-INL**
--------
Pull Request Description
--------
##### What issue does this change request address?
#245 
#248 
#250 

##### What are the significant changes in functionality due to this change request?
1. HERON dependencies have been updated to pip install newest DISPATCHES version (still as an optional dependency)
2. HERD restructured the build function that creates the Pyomo Multiperiod model to match the newest DISPATCHES release. It now creates a Multiperiod IDAES object that can handle stochastic sets. 
3. Python imports of the DISPATCHES package match the newest re-organization of the DISPATCHES repository.
4. HERD now uses an intermediate `flowsheet_block` method to avoid redundant creation of Pyomo Models; instead it creates one model per realization/scenario and for each (time, cluster, year) timestep within that realization it clones from the original. 
5. Fixed error involving wrong NPV values when `num_samples` > 1. Regolded `nuclearCase_Sine` test because that was the only test with more than 1 scenario. 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [x] 4. Automated Tests should pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [x] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Review by PaulTalbot-INL — APPROVED** [review]
LGTM

File: src/Herd.py [hunk 1]
```diff
@@ -1037,11 +1139,12 @@ def _add_objective(self):
...
     """
-    # scenario probability weights
-    weights = self._dmdl.weights_scenarios
+    ## scenario probability weights
+    # weights = self._time_sets['weights_scenarios'] # TODO: skipping for now
 
     # pyomo expression for full metric wtih scenario weights applied
-    Metric = np.sum( weights[n]*scenario['NPV'] for n, scenario in enumerate(self._metrics) )
+    N = getattr(self, "_num_samples")
+    Metric = self._metrics[-1]['NPV'] / N # TODO: temp fix, things are getting duplicated somewhere
```
Comment (PaulTalbot-INL) [inline src/Herd.py:1147]: are the weights being correctly normalized to add to 1? Glad this works, curious about what's off.

**Comment by PaulTalbot-INL** [comment]
The required tests have passed, checklist clears, merging.

File: src/Herd.py (hunk 1 above)
Comment (GabrielSoto-INL) [reply src/Herd.py:1147]: the weights are normalized to 1, the problem is in the loop within `_build_dispatches_model`. HERD loops through all scenarios/realizations and creates an NPV metric for each. But by the time it gets to the final scenario, it keeps cash flows from previous scenarios. This is a temp fix to make sure all cash flows are accounted for, I plan to dig a little deeper and implement a nicer solution.

Comment (PaulTalbot-INL) [reply src/Herd.py:1147]: Oh, I see, interesting! Thanks.

**Pull request merged**